cmd/thanos/query.go

```diff
@@ -690,7 +690,15 @@ func runQuery(
 			grpcserver.WithMaxConnAge(grpcMaxConnAge),
 		)
 
+		firstServiceDiscovery := true
+
 		g.Add(func() error {
+			if firstServiceDiscovery {
+				ctx, cancel := context.WithCancel(context.Background())
+				endpoints.Update(ctx)
```

Comment:
The problem is that the endpoints will be updated twice on boot. :( Why not use channels for this? Something like:

```go
initialDiscovery := make(chan int)

// Add this where the update happens.
g.Add(func() error {
   ...
   endpoints.Update(ctx)
   initialDiscovery <- 1
   close(initialDiscovery)
   ...
}, ...)

// Then add this just before statusProber.Ready()
<-initialDiscovery
```

That should do it, right?

Here's an example: https://go.dev/play/p/-vKlpjXKpCk.

Reply:
I thought about doing this only, the problem is we are updating our endpoints every 5 seconds, take a look at the function where we are updating in query.go and my comments,

```
// Periodically update the store set with the addresses we see in our cluster.
	{
		ctx, cancel := context.WithCancel(context.Background())
		g.Add(func() error {
			return runutil.Repeat(5*time.Second, ctx.Done(), func() error { 
// if we write a value to a channel inside this function and close it afterwards, means we actually do it every 5 seconds, and obviously a channel cannot be closed more than once, will lead to a deadlock
				endpoints.Update(ctx)
				return nil
			})
		}, func(error) {
			cancel()
			endpoints.Close()
		})
	}
```
Moreover, if we write something to the channel it will block that goroutine until some other goroutine reads it from the channel, Here we will read the value from the channel only once, since we do statusProber.Ready() only once, while we write in it after every 5 seconds, which again leads to a deadlock

Reply:
So, maybe we could use a variable there to avoid trying to close the channel twice? :thinking: We would still avoid the problem of calling Update() twice on startup.

Reply:
Or how about 1) do and wait on the first service discovery; 2) once the first service discovery is run, only then we kick off the periodical updates logic and unblock server from starting to listen. Does that make sense?